File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-0.py


"""

[3,4,5,6,1,2]
       l t r
         m

m > l
we are in the left side
our target is between l and m?

[4,5,1,2,3]
 l t m   r
 l r
 m
 t
m < l
we are in the right side
our target is between m and r?

if nums[m] == target:
    return m

if nums[m] >= nums[l]:
    if target > nums[m] or target < nums[l]:
        l = m + 1
    else:
        r = m - 1
elif nums[m] < nums[l]:
    if target < nums[m] or target > nums[r]:
        r = m - 1
    else:
        l = m +1
        


"""

class Solution:
    def search(self, nums: List[int], target: int) -> int:  
        l = 0
        r = len(nums) - 1
        res = -1
        while l <= r:
            m = (l+r)//2

            if nums[m] == target:
                res = m
                break
            
            if nums[m] >= nums[l]:
                if target > nums[m] or target < nums[l]:
                    l = m + 1
                else:
                    r = m - 1
            else:
                if target < nums[m] or target > nums[r]:
                    r = m - 1
                else:
                    l = m + 1
        
        return res

        

File: Data Structures & Algorithms/max-water-container/submission-0.cs
/*
 - we could apply the two pointers approach
 - two pointers 
    - left
    - right

 - Calculate the volume while we decrease both index pointers
 - Saving always the maximum volume found it and the height indexes

 - Icrement or decrement the smaller height of the index side
 - Decrement the width between the pointers

 - [1,7,2,5,4,7,3,6]
    ^             ^
    |_____________|
           7

    volume = Min(1(left),6(right)) * 7 (width) = 7
    maximum = 7
 - [1,7,2,5,4,7,3,6]
      ^           ^
      |___________|
           6

    volume = Min(7(left),6(right)) * 6 (width) = 36
    maximum = 36

 - [1,7,2,5,4,7,3,6]
      ^         ^
      |_________|
           5

    volume = Min(7(left),3(right)) * 5 (width) = 15
    maximum = 36
    
 - [1,7,2,5,4,7,3,6]
        ^     ^
        |_____|
           3

    volume = Min(2(left),7(right)) * 3 (width) = 6
    maximum = 36
*/
public class Solution {
    public int MaxArea(int[] heights) {
        int left = 0, right = heights.Length -1;
        int width = heights.Length - 1;
        int maxVolume = 0;

        while(left < right){
            int currentVolume = Math.Min(heights[left], heights[right]) * width;

            if(currentVolume > maxVolume)
                maxVolume = currentVolume;
            
            if(heights[left] >= heights[right]){
                right--;
            } else {
                left++;
            }

            width--;
        }

        return maxVolume;
    }
}
